extensions/oidc-client/runtime/src/main/java/io/quarkus/oidc/client/runtime/OidcClientImpl.java

```diff
@@ -0,0 +1,150 @@
+    public Uni<GrantTokens> getTokens() {
+        return getJsonResponse(tokenGrantParams, false);
+    }
+
+    @Override
+    public Uni<GrantTokens> refreshTokens(String refreshToken) {
+        if (refreshToken == null) {
+            throw new OidcClientException("Refresh token is null");
+        }
+        MultiMap refreshGrantParams = copyMultiMap(commonRefreshGrantParams);
+        refreshGrantParams.add(REFRESH_TOKEN, refreshToken);
+        return getJsonResponse(refreshGrantParams, true);
+    }
+
+    private Uni<GrantTokens> getJsonResponse(MultiMap body, boolean refresh) {
+        HttpRequest<Buffer> request = client.post(tokenRequestUri);
+        if (clientSecretBasicAuthScheme != null) {
+            request.putHeader(AUTHORIZATION_HEADER, clientSecretBasicAuthScheme);
+        } else if (clientJwtKey != null) {
+            // if it is a refresh then a map has already been copied
+            body = !refresh ? copyMultiMap(body) : body;
+            body.add(CLIENT_ASSERTION_TYPE, JWT_BEARER_CLIENT_ASSERTION_TYPE);
+            body.add(CLIENT_ASSERTION, OidcCommonUtils.signJwtWithKey(oidcConfig, clientJwtKey));
+        }
+        return request.sendForm(body).onItem()
+                .transformToUni(resp -> emitGrantTokens(resp, refresh));
+    }
+
+    private Uni<GrantTokens> emitGrantTokens(HttpResponse<Buffer> resp, boolean refresh) {
+        return Uni.createFrom().emitter(new Consumer<UniEmitter<? super GrantTokens>>() {
+            @Override
+            public void accept(UniEmitter<? super GrantTokens> emitter) {
+                if (resp.statusCode() == 200) {
+                    LOG.debugf("Tokens have been %s", refresh ? "refreshed" : "acquired");
+                    JsonObject json = resp.bodyAsJsonObject();
+                    final String accessToken = json.getString(ACCESS_TOKEN);
+                    final String refreshToken = json.getString(REFRESH_TOKEN);
+                    Long accessTokenExpiresAt = json.getLong(EXPIRES_AT);
+                    if (accessTokenExpiresAt == null) {
+                        accessTokenExpiresAt = getExpiresJwtClaim(accessToken);
+                    }
+                    emitter.complete(new GrantTokens(accessToken, accessTokenExpiresAt, refreshToken));
+                } else {
+                    LOG.errorf("%s token grant request has failed: %s", (refresh ? "refresh" : grantType), resp.bodyAsString());
+                    emitter.fail(new OidcClientException());
+                }
+            }
+        });
+    }
+
+    private static Long getExpiresJwtClaim(String accessToken) {
+        String[] parts = accessToken.split("\\.");
+        if (parts.length == 3) {
+            try {
+                JsonObject claims = new JsonObject(new String(Base64.getUrlDecoder().decode(parts[1]), StandardCharsets.UTF_8));
+                return claims.getLong(Claims.exp.name());
+            } catch (IllegalArgumentException ex) {
+                return null;
```

Comment:
This still needs to be addressed.

Reply:
@stuartwdouglas sorry, missed it. OK, a token grant response should include an expiry time, but if it is not included, (as in case of Keycloak) then an attempt is made to decode the access token as JWT token and get the exp claim - this exception is only to handle the `JsonObject.getLong` exception - should not ever happen in practice - but indeed, some debug message should be added